src/compiler/checker.ts

```diff
@@ -11504,60 +11509,72 @@ namespace ts {
             return resultType;
 
             function getTypeAtFlowNode(flow: FlowNode): FlowType {
+                const saveFlowLength = flowLength;
                 while (true) {
-                    if (flow.flags & FlowFlags.Shared) {
+                    flowLength++;
+                    if (flowLength === 5000) {
+                        // The length of this particular control flow path is 5000 nodes or more. Rather than spending an
+                        // excessive amount of time and possibly overflowing the call stack, we report an error and disable
+                        // further control flow analysis in the containing function or module body.
+                        flowAnalysisDisabled = true;
+                        error(reference, Diagnostics.The_body_of_the_containing_function_or_module_is_too_large_for_control_flow_analysis);

```

Comment:
we talked about putting this on the containing function if one exists. this way ppl have a consistent place to see the error, instead of it jumping around as they add/remove lines.